mlir/docs/Canonicalization.md

```diff
@@ -275,3 +275,26 @@ Operation *MyDialect::materializeConstant(OpBuilder &builder, Attribute value,
   ...
 }
 ```
+
+### Choosing between implementing a Folder or a `RewritePattern`
+
+When implementing a new canonicalization, an important thing to think about is
+if the canonicalization should be a folder or a `RewritePattern`. The generally
+accepted policy is:
+
+*   If a transformation should be a canonicalizer is a different question from
+    if a transformation should be a folder or a `RewritePattern`. A
+    transformation is promoted to a canonicalization as defined by the General
+    Design of canonicalizations.
```

Comment:
Maybe write this in a more imperative way? For example, you could add a 'NOTE' marker that says: 'To decide whether a transformation should be a canonicalization in the first place, see the General Design of Canonicalization.' (link)

Reply:
I'm not sure if I want it to be a NOTE. I'm trying to stress the point here that you cannot say that a transformation should be a RewritePattern canonicalization, but it shouldn't be a folder canonicalization based on what the transformation does. It's an implementation detail. The only thing you have to reason about is if a transformation should be a canonicalization or not. If it's a folder or a RewritePattern is dependent on it's implementation.